platform/mds/mds/src/main/java/org/motechproject/mds/service/impl/csv/writer/PdfTableWriter.java

```diff
@@ -108,15 +194,35 @@ private float calculateTotalTableWidth(float[] widths) {
         return totalWidth;
     }
 
-    private void updateWidthIfNeeded(String header, PdfPCell cell) {
-        Float width = calculateCellWidth(cell);
+    private float calculateTotalColumnsWidth(float[] relativeWidths, int firstColumn, int lastColumn) {
+        float totalColumnsSize = 0;
 
-        if (columnsWidths.get(header) < width) {
-            columnsWidths.put(header, width > MAX_COLUMN_WIDTH ? MAX_COLUMN_WIDTH : width);
+        for (int column = firstColumn; column < lastColumn; column++) {
+            totalColumnsSize += relativeWidths[column];
+        }
+
+        return totalColumnsSize;
+    }
+
+    private List<Integer> calculateLastColumnsForPages(float[] relativeWidths) {
+        List<Integer> lastColumnsForPages = new LinkedList<>();
```

Comment:
Why LinkedList? I can see you use get by index on the list, which for LinkedList has complexity O(n), while for ArrayList this is obviously O(1). Unless you plan to access/delete elements with iterators use ArrayLists